test/jdk/tools/jpackage/macosx/SigningAppImageTwoStepsTest.java

```diff
@@ -0,0 +1,124 @@
+ * launcher and app bundle itself. This test requires that machine is configured with
+ * test certificate for "Developer ID Application: jpackage.openjdk.java.net" or
+ * alternately "Developer ID Application: " + name specified by system property:
+ * "jpackage.mac.signing.key.user.name" in the jpackagerTest keychain
+ * (or alternately the keychain specified with the system property
+ * "jpackage.mac.signing.keychain". If this certificate is self-signed, it must
+ * have be set to always allowed access to this keychain" for user which runs test.
+ * (If cert is real (not self signed), the do not set trust to allow.)
+ */
+
+/*
+ * @test
+ * @summary jpackage with --type app-image --app-image "appImage" --mac-sign
+ * @library ../helpers
+ * @library /test/lib
+ * @library base
+ * @build SigningBase
+ * @build SigningCheck
+ * @build jtreg.SkippedException
+ * @build jdk.jpackage.test.*
+ * @build SigningAppImageTwoStepsTest
+ * @modules jdk.jpackage/jdk.jpackage.internal
+ * @requires (os.family == "mac")
+ * @run main/othervm/timeout=360 -Xmx512m jdk.jpackage.test.Main
+ *  --jpt-run=SigningAppImageTwoStepsTest
+ */
+public class SigningAppImageTwoStepsTest {
+
+    @Test
+    @Parameter("true")
+    @Parameter("false")
+    public void test(boolean signAppImage) throws Exception {
+        SigningCheck.checkCertificates();
+
+        Path appimageOutput = TKit.workDir().resolve("appimage");
+
+        // Generate app image. Signed or unsigned based on test
+        // parameter. We should able to sign predfined app images
+        // which are signed or unsigned.
+        JPackageCommand appImageCmd = JPackageCommand.helloAppImage()
+                .setArgumentValue("--dest", appimageOutput);
+        if (signAppImage) {
+            appImageCmd.addArguments("--mac-sign", "--mac-signing-key-user-name",
+                    SigningBase.DEV_NAME, "--mac-signing-keychain",
+                    SigningBase.KEYCHAIN);
+        }
+
+        // Add addtional launcher
+        AdditionalLauncher testAL = new AdditionalLauncher("testAL");
+        testAL.applyTo(appImageCmd);
+
+        // Generate app image
+        appImageCmd.executeAndAssertHelloAppImageCreated();
+
+        // Double check if it is signed or unsigned based on signAppImage
+        verifySignature(appImageCmd, signAppImage);
+
+        // Sign app image
+        JPackageCommand cmd = JPackageCommand.helloAppImage()
+                .addArguments("--app-image", appImageCmd.outputBundle().toAbsolutePath())
+                .addArguments("--mac-sign")
+                .addArguments("--mac-signing-key-user-name", SigningBase.DEV_NAME)
+                .addArguments("--mac-signing-keychain", SigningBase.KEYCHAIN)
+                .removeArgumentWithValue("--input")
+                .removeArgumentWithValue("--dest")
+                .removeArgumentWithValue("--name")
+                .removeArgumentWithValue("--main-jar")
+                .removeArgumentWithValue("--main-class")
```

Comment:
There is no point in calling `JPackageCommand.helloAppImage()` if later all arguments added by this call get deleted. Simply call JPackageCommand ctor instead.